### PR TITLE
Support configurable LangChain4j A2A agents

### LINKED ISSUE #132: Support configurable LangChain4j A2A agents
* Extension: LangChain4j
* Extension Version: 27.0.0-SNAPSHOT

## Task Description

Add optional support for LangChain4j A2A client agents without affecting applications that do not use A2A.

The integration should:

- Expose the optional `langchain4j-agentic-a2a` module through the extension BOM.
- Support `a2a-server-url`, `output-key`, and `async` overrides under `langchain4j.agents.<agent-name>`.
- Apply configuration over annotation defaults while preserving existing `@A2AClientAgent` behavior.
- Support both top-level `@Ai.Agent` services and A2A agents nested in composed workflows.
- Provide clear validation for missing, conflicting, or invalid configuration.
- Include credential-free loopback A2A contract tests, code-generation tests, negative tests, and documentation.
- Run deterministically in the existing Linux and Windows PR test matrix without external services.

## Configuration Override Example

A remote agent may declare defaults with `@A2AClientAgent`:

```java
@Ai.Agent("remote-writer")
public interface RemoteWriter {

    @A2AClientAgent(
            a2aServerUrl = "https://writer.example.com",
            outputKey = "draft",
            async = true)
    String write(@V("topic") String topic);
}
```

Helidon configuration overrides those annotation values by matching the `@Ai.Agent` name:

```yaml
langchain4j:
  agents:
    remote-writer:
      a2a-server-url: https://writer.internal.example/a2a
      output-key: story
      async: false
```

The resulting agent uses `https://writer.internal.example/a2a`, writes its result under `story`, and executes
synchronously (`async: false`).

The implementation PR will also include credential-free HTTP contract tests for the existing OpenAI, Gemini, Cohere, and Ollama providers as its first commit.

### PR DESCRIPTION
**Pull request by danielkec**
### Description

Resolves #132

- Add credential-free HTTP contract tests for the OpenAI, Gemini, Cohere, and Ollama providers as the first commit.
- Add `a2a-server-url`, `output-key`, and `async` overrides for LangChain4j A2A client agents.
- Support generated top-level and nested A2A agents without requiring a local chat model.
- Preserve annotation and supplier fallbacks, client customizers, listeners, composed-agent precedence, and upstream behavior for unannotated A2A subagents.
- Manage the optional `langchain4j-agentic-a2a` dependency in the extension BOM.
- Add unit/code-generation coverage and a credential-free loopback Agent Card/JSON-RPC contract module.

### Documentation

- Document remote A2A agents, configuration precedence, the Agent Card customizer limitation, and the upstream A2A SDK module-path split-package limitation.
- Add `a2a-server-url` to the generated configuration reference.

### Testing

- `bash ./etc/scripts/copyright.sh`
- `bash ./etc/scripts/checkstyle.sh`
- Focused A2A reactor: 11 runtime/code-generation tests and one loopback contract test.
- Existing agentic and agentic-MCP regression selection: 4 agentic and 4 MCP tests.
- Full 21-module `-Pspotbugs,javadoc,tests,examples -DskipTests verify`.
- Provider contract test reactors and the full LangChain4j reactor were validated for the first commit.
- The complete GitHub PR matrix passed on Linux and Windows, including all nine required checks.

### CODE REVIEW

File: extensions/langchain4j/modules/langchain4j/src/main/java/io/helidon/extensions/langchain4j/A2AAgentConfigSupport.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+    /**
+     * Creates an A2A client agent, applying Helidon configuration over annotation defaults.
+     *
+     * @param agentType agent service type
+     * @param config agent configuration
+     * @param <T> agent service type
+     * @return configured A2A client agent
+     */
+    public static <T> T create(Class<T> agentType, AgentsConfig config) {
```
Comment (tomas-langer) [inline extensions/langchain4j/modules/langchain4j/src/main/java/io/helidon/extensions/langchain4j/A2AAgentConfigSupport.java]: AI Finding: `AgentsConfig.enabled()` documents that `false` makes the agent unavailable, but both generated A2A call sites invoke this method unconditionally and this method never checks `config.enabled()` before constructing the client. For a configuration-only A2A endpoint, resolving the service or a containing workflow therefore still performs Agent Card discovery—and later permits prompt traffic—even when the agent is explicitly disabled. This defeats the documented operational kill switch and can preserve unwanted outbound network and data-disclosure paths.

File: extensions/langchain4j/modules/langchain4j/src/main/java/io/helidon/extensions/langchain4j/A2AAgentConfigSupport.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+import static dev.langchain4j.agentic.declarative.DeclarativeUtil.checkReturnType;
+import static dev.langchain4j.agentic.declarative.DeclarativeUtil.invokeStatic;
+import static dev.langchain4j.agentic.declarative.DeclarativeUtil.selectMethod;
+import static dev.langchain4j.agentic.internal.AgentUtil.getAnnotatedMethodOnClass;
+
+/**
+ * Runtime support for applying Helidon configuration to LangChain4j A2A client agents.
+ */
+@Api.Internal
+public final class A2AAgentConfigSupport {
```
Comment (tomas-langer) [inline extensions/langchain4j/modules/langchain4j/src/main/java/io/helidon/extensions/langchain4j/A2AAgentConfigSupport.java]: AI Finding: This new public `@Api.Internal` class sits in the exported `io.helidon.extensions.langchain4j` package, and generated application code compiles against its public methods. Rule 4.1.5 treats every public class and method as maintained Helidon API, so this makes runtime/codegen plumbing part of the consumer-visible compatibility surface and couples future generated code to LangChain4j internal APIs.

File: extensions/langchain4j/docs/config/io.helidon.extensions.langchain4j.AgentsConfig.md [hunk 3]
```diff
@@ -19,6 +19,17 @@ Configuration for a single LangChain4j agent
...
 <tr>
 <td>
+<code>a2a-<wbr>server-<wbr>url</code>
+</td>
+<td>
+<code>String</code>
+</td>
+<td>
+</td>
+<td>URL of the remote A2A agent server</td>
```
Comment (tomas-langer) [inline extensions/langchain4j/docs/config/io.helidon.extensions.langchain4j.AgentsConfig.md]: AI Finding: The new config reference describes this value only as a URL, but `A2AAgentConfigSupport` accepts only absolute HTTP(S) URIs with a host and rejects blank values, user info, fragments, and malformed or oversized ports. Plausible configurations therefore fail at startup even though neither this generated reference nor the authored A2A guide describes the accepted URI contract.

File: extensions/langchain4j/docs/langchain4j.md [hunk 4]
```diff
@@ -453,6 +453,65 @@ entries, for example to replace a chat model or adjust an output key.
...
+```
+
+The A2A URL, output key, and asynchronous execution setting can be overridden
+without recompiling the application:
+
+```yaml
+langchain4j:
+  agents:
+    remote-writer:
+      a2a-server-url: http://writer.internal:8080
```
Comment (tomas-langer) [inline extensions/langchain4j/docs/langchain4j.md]: AI Finding: This user-facing example configures a non-loopback remote agent over cleartext HTTP, so Agent Card discovery and subsequent prompt/result traffic can traverse the network unencrypted. The following note also states that `@A2AClientCustomizer` cannot add TLS to the discovery request, making the example unsafe to copy into the internal-service scenario it depicts without any development-only qualification.

File: extensions/langchain4j/modules/codegen/src/main/java/io/helidon/extensions/langchain4j/codegen/AgentCodegen.java [hunk 5]
```diff
@@ -54,6 +57,14 @@
 
 class AgentCodegen implements CodegenExtension {
     private static final TypeName GENERATOR = TypeName.create(AgentCodegen.class);
+    private static final Set<TypeName> COMPOSED_AGENT_ANNOTATIONS = Set.of(
```
Comment (tomas-langer) [inline extensions/langchain4j/modules/codegen/src/main/java/io/helidon/extensions/langchain4j/codegen/AgentCodegen.java]: AI Finding: This new private static final field is declared before the package-private static final `AGENTS_CONFIG_KEY` at line 68. Helidon Development Guideline Rule 2.1 requires static final fields to be ordered by visibility, with package-private fields before private fields, so the changed class violates the project’s required member order.

File: extensions/langchain4j/modules/providers/open-ai/src/test/java/io/helidon/extensions/langchain4j/providers/openai/OpenAiProviderContractTest.java [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+            EmbeddingModel embeddingModel = registry.getNamed(EmbeddingModel.class, EMBEDDING_MODEL_NAME);
+            assertThat(embeddingModel, instanceOf(OpenAiEmbeddingModel.class));
+
+            ContractChatService chatService = registry.get(ContractChatService.class);
+            assertThat(chatService.chat("sync contract prompt"), is("openai-sync-ok"));
+
+            ContractStreamingService streamingService = registry.get(ContractStreamingService.class);
+            assertThat(streamingService.chat("stream contract prompt").collect(Collectors.joining()),
+                       is("openai-stream-ok"));
+            assertArrayEquals(new float[] {0.125F, -0.5F, 0.75F},
```
Comment (tomas-langer) [inline extensions/langchain4j/modules/providers/open-ai/src/test/java/io/helidon/extensions/langchain4j/providers/openai/OpenAiProviderContractTest.java]: AI Finding: This new test uses JUnit `assertArrayEquals`, and the same change also introduces JUnit `assertTrue` at line 157 and `assertArrayEquals` in `OllamaProviderContractTest`. Helidon Development Guideline Rules 10.5 and 10.6 require Hamcrest `assertThat(...)` assertions and limit JUnit assertions to `assertAll` and `assertThrows`, so these new contract tests violate the project’s assertion policy and provide inconsistent diagnostics.

File: extensions/langchain4j/modules/langchain4j/src/main/java/io/helidon/extensions/langchain4j/AgentsConfigBlueprint.java [hunk 7]
```diff
@@ -71,6 +72,19 @@ interface AgentsConfigBlueprint {
     @Option.Configured
     Optional<String> description();
 
+    /**
+     * URL of the remote A2A agent server.
+     * <p>
+     * When configured for an A2A client agent, this value overrides the URL declared by
+     * {@code A2AClientAgent} or supplied by {@code A2AServerUrlSupplier}.
+     *
+    * @return configured A2A server URL, or empty if the annotation configuration should be used
```
Comment (tomas-langer) [inline extensions/langchain4j/modules/langchain4j/src/main/java/io/helidon/extensions/langchain4j/AgentsConfigBlueprint.java]: AI Finding: The added `@return` line places the leading `*` one column earlier than every other line in the block. This malformed indentation is a changed-source style defect in the new public configuration accessor.

File: extensions/langchain4j/modules/langchain4j/src/main/java/io/helidon/extensions/langchain4j/AgentsConfigSupport.java [hunk 8]
```diff
@@ -38,6 +38,37 @@ class AgentsConfigSupport {
...
+        return A2AAgentConfigSupport.create(agentType, agentsConfig);
+    }
+
+    /**
+     * Whether the type declares an A2A client agent method and no higher-precedence nested-agent method.
+     *
+     * @param agentType agent service type
+     * @return whether the type should use A2A client agent creation as a nested agent
+     */
+    @Prototype.PrototypeFactoryMethod
```
Comment (tomas-langer) [inline extensions/langchain4j/modules/langchain4j/src/main/java/io/helidon/extensions/langchain4j/AgentsConfigSupport.java]: AI Finding: `@Prototype.PrototypeFactoryMethod` copies this helper as public static `AgentsConfig.isA2ASubAgent(Class<?>)`; the builder generator forces custom factory methods public, and `io.helidon.extensions.langchain4j` is exported. The only production call is from the generated provider's private `resolveSubAgent` path, so this internal LangChain4j dispatch classifier becomes consumer-visible maintained API under [Rule 4.1.5](https://github.com/helidon-io/helidon/blob/main/DEV-GUIDELINES.md#rule-4-1-5). Future changes to composed-agent precedence would be constrained by a public method that is not part of the configuration contract.

File: extensions/langchain4j/modules/langchain4j/src/main/java/io/helidon/extensions/langchain4j/AgentsConfigSupport.java [hunk 9]
```diff
@@ -38,6 +38,37 @@ class AgentsConfigSupport {
...
+     * @param agentsConfig agent configuration
+     * @param agentType agent service type
+     * @param <T> agent service type
+     * @return configured A2A client agent
+     * @throws IllegalArgumentException if {@code agentType} is not an A2A client agent, its server URL sources are
+     *                                  missing or conflicting, its configured URL or output-key override is invalid, or
+     *                                  annotated method metadata is invalid
+     */
+    @Prototype.PrototypeMethod
+    static <T> T createA2AAgent(AgentsConfig agentsConfig, Class<T> agentType) {
```
Comment (tomas-langer) [inline extensions/langchain4j/modules/langchain4j/src/main/java/io/helidon/extensions/langchain4j/AgentsConfigSupport.java:58]: AI Finding: Both new custom methods (`createA2AAgent` here and `isA2ASubAgent` below) are generated as public `AgentsConfig` API and accept `Class<?> agentType`, but neither validates that argument before passing it to internal reflection helpers. [Rule 1.2.2](https://github.com/helidon-io/helidon/blob/main/DEV-GUIDELINES.md#rule-1-2-2) requires non-null public API parameters to be validated at the boundary. A null argument therefore fails at an implementation-dependent downstream dereference rather than at the new public API boundary, making the public failure contract unpredictable.

File: extensions/langchain4j/modules/langchain4j/src/test/java/io/helidon/extensions/langchain4j/A2AAgentConfigSupportTest.java [hunk 10]
```diff
@@ -0,0 +1,497 @@
...
+            var composedType = Class.forName(getClass().getPackageName()
+                                                     + ".A2AAgentConfigSupportTest_ComposedAgent__AiAgent");
+            var composedConstructor = composedType.getDeclaredConstructor(Config.class,
+                                                                          ServiceRegistry.class,
+                                                                          Optional.class);
+            composedConstructor.setAccessible(true);
+            var composedSupplier = composedConstructor.newInstance(invalidConfig,
+                                                                   manager.registry(),
+                                                                   Optional.empty());
+            var resolver = composedType.getDeclaredMethod("resolveSubAgent", Class.class);
```
Comment (tomas-langer) [inline extensions/langchain4j/modules/langchain4j/src/test/java/io/helidon/extensions/langchain4j/A2AAgentConfigSupportTest.java]: AI Finding: These new tests load generated classes by constructed names, assert constructor signatures, and use `setAccessible(true)` to invoke private `resolveSubAgent` methods, including the disabled-nested and unannotated-fallback paths below. [Rules 10.1 and 10.1.1](https://github.com/helidon-io/helidon/blob/main/DEV-GUIDELINES.md#rule-10-1) require tests to cover observable behavior unless structure itself is the contract, and [Rule 10.2](https://github.com/helidon-io/helidon/blob/main/DEV-GUIDELINES.md#rule-10-2) prohibits reflection into private implementation state. These assertions bind coverage to private generated names, access levels, constructor shapes, and helper methods, so non-behavioral codegen refactors will break the tests even when the public `AgentsConfig` and registry behavior is unchanged.

File: extensions/langchain4j/modules/codegen/src/main/java/io/helidon/extensions/langchain4j/codegen/AgentCodegen.java [hunk 11]
```diff
@@ -109,58 +120,69 @@ private void process(RoundContext roundCtx, TypeInfo agentInterface) {
...
+                        .addContent(".class, ")
+                        .addContent("configuredModel, new ")
+                        .addContent(LC_AGENTIC_SERVICES)
+                        .addContent(".AgentConfigurator(this")
+                        .addContentLine("::configureSubAgents, this::resolveSubAgent, null));")
+                        .addContentLine("");
+            }
+        });
 
+        classModel.addMethod(this::addAgentsConfigMethod);
```
Comment (tomas-langer) [inline extensions/langchain4j/modules/codegen/src/main/java/io/helidon/extensions/langchain4j/codegen/AgentCodegen.java]: AI Finding: `addAgentsConfigMethod` emits a private instance method, but it is added before the existing package-private `configureSubAgents` method on the next line. `ClassModel` preserves this insertion order, so generated `__AiAgent` classes place a private instance method before a non-private instance method. [Rule 2.1](https://github.com/helidon-io/helidon/blob/main/DEV-GUIDELINES.md#rule-2-1) requires non-private instance methods to precede private instance methods, making every generated class from this code path violate the required member order.

File: extensions/langchain4j/modules/langchain4j/src/test/java/io/helidon/extensions/langchain4j/A2AAgentConfigSupportTest.java [hunk 12]
```diff
@@ -0,0 +1,497 @@
...
+                """;
+        return Config.just(ConfigSources.create(yaml, MediaTypes.APPLICATION_X_YAML));
+    }
+
+    @Ai.Agent("annotated-a2a")
+    public interface AnnotatedAgent {
+        @A2AClientAgent(a2aServerUrl = "a2a+test:annotation",
+                        outputKey = "annotation-output",
+                        async = true)
+        String ask(@V("question") String question);
```
Comment (tomas-langer) [inline extensions/langchain4j/modules/langchain4j/src/test/java/io/helidon/extensions/langchain4j/A2AAgentConfigSupportTest.java:325]: AI Finding: The new A2A fixture interfaces declare instance agent methods before non-private static supplier or customizer methods, including `SuppliedAgent` and `ConflictingAgent` below and `UnannotatedRemoteWriterAgent` in the contract-test module. [Rule 2.1](https://github.com/helidon-io/helidon/blob/main/DEV-GUIDELINES.md#rule-2-1) requires non-private static methods to precede non-private instance methods. These new sources therefore violate the required member ordering at each of those declarations.

File: extensions/langchain4j/modules/codegen/src/main/java/io/helidon/extensions/langchain4j/codegen/AgentCodegen.java [hunk 13]
```diff
@@ -243,20 +268,106 @@ private void addConfigureSubAgentsMethod(Method.Builder mb) {
...
+                && COMPOSED_AGENT_ANNOTATIONS.stream().noneMatch(annotation -> hasAnnotation(typeInfo, annotation));
+    }
+
+    private boolean hasAnnotation(TypeInfo typeInfo, TypeName annotation) {
+        if (typeInfo.elementInfo()
+                .stream()
+                .anyMatch(element -> element.hasAnnotation(annotation))) {
+            return true;
+        }
+        return typeInfo.interfaceTypeInfo()
```
Comment (tvallin) [inline extensions/langchain4j/modules/codegen/src/main/java/io/helidon/extensions/langchain4j/codegen/AgentCodegen.java]: [P2] This recursion does not account for method overriding. If a child interface redeclares an annotated parent method without `@A2AClientAgent`, this still returns true, while runtime `getAnnotatedMethodOnClass` uses `Class.getMethods()` and sees only the unannotated child declaration. The generated supplier is then classified as A2A and omits `ChatModel`, but `createA2AAgent` rejects it at runtime. Please derive effective methods using override-shadowing semantics.

**Review by tvallin — APPROVED** [review]
